### PR TITLE
[NewPM][AMDGPU] Port SIPreAllocateWWMRegs to NPM

### PR DESCRIPTION
**Pull request by optimisan**

### CODE REVIEW

**Comment by Akshat-Oke** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109939?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#109939** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109939?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#109938** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109938?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109937** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109937?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#109936** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/109936?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Akshat-Oke and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp [hunk 1]
```diff
@@ -254,3 +262,13 @@ bool SIPreAllocateWWMRegs::runOnMachineFunction(MachineFunction &MF) {
   rewriteRegs(MF);
   return true;
 }
+
+PreservedAnalyses
+SIPreAllocateWWMRegsPass::run(MachineFunction &MF,
+                              MachineFunctionAnalysisManager &MFAM) {
+  auto *LIS = &MFAM.getResult<LiveIntervalsAnalysis>(MF);
+  auto *Matrix = &MFAM.getResult<LiveRegMatrixAnalysis>(MF);
+  auto *VRM = &MFAM.getResult<VirtRegMapAnalysis>(MF);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:271]: I thought this should be using getCachedResult, and the pass supported no LIS for the fast RA path. But I see now the legacy path is requiring them (although it probably shouldn't?)

File: llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
+//===--- SIPreAllocateWWMRegs.h -------------------------------------------===//
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.h:1]: Missing C++ mode comment

File: llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp (hunk 1 above)
Comment (cdevadas) [reply llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:271]: Yep, this looks odd. We shouldn't involve VRM for the fast regalloc case.

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Comment by optimisan** [comment]
### Merge activity

* **Oct 22, 5:41 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/109939).

**Pull request merged**